torch/utils/data/dataset.py

```diff
@@ -260,6 +265,125 @@ def __len__(self):
+    (or simply the size) is unknown. Only the number of chunks is known for the dataset.
+
+    In a distributed setup, each `DataLoader` worker has an instance of ``ChunkDataset`` that uses
+    ``DistributedChunkSampler`` sampler to select chunks based on their global worker rank.
+    Once a chunk index has been selected, it is passed to ``ChunkDataReader`` that will load
+    data into the ``ChunkDataset`` internal cache. Only then ``ChunkDataset`` returns batches.
+
+    Arguments:
+        chunk_sampler (DistributedSampler or DistributedChunkSampler): Draw indices for chunks.
+            Typically used to split data amongst dataloader workers (first level of sampling)
+        chunk_reader (ChunkDataReader): Specialized reader for a given input type
+        shuffle_cache (bool): Setting `True` trigger shuffling of the internal chunk cache
+            (second level of sampling) (default: `True`)
+    """
+
+    def __init__(self, chunk_sampler, chunk_reader, shuffle_cache=True):
+        super(ChunkDataset, self).__init__()
+        assert isinstance(chunk_sampler, DistributedSampler) or isinstance(
+            chunk_sampler, DistributedChunkSampler), 'sampler must be a `DistributedSampler` or `DistributedChunkSampler`'
+        assert callable(chunk_reader), 'chunk_reader must be `callable()` and return a container with data'
+        assert isinstance(shuffle_cache, bool), 'shuffle_cache must be a `bool`'
+
+        self.chunk_sampler = chunk_sampler
+        self.chunk_reader = chunk_reader
+        self.shuffle_cache = shuffle_cache
+
+        # Internal state
+        self._chunk_sampler_iter = iter(self.chunk_sampler)
+        self._cache = []
+        self._batch_size = 1
+        self._min_cache = 1000
+
+    def __iter__(self):
+        return self
+
+    def __len__(self):
+        raise NotImplementedError
+
+    def __next__(self):
+        r"""Returns a batch or raises exception when exhausted"""
+
+        cache_size = 0
+        if len(self._cache) > 0:
+            cache_size = len(self._cache)
+
+        if (cache_size < self._min_cache):
+            new_chunk = None
+            # Try reading a new chunk
+            # Ignore EOF here because there may be internal cache entries to be returned yet
+            try:
+                new_chunk = self.chunk_reader(next(self._chunk_sampler_iter))
+                assert isinstance(new_chunk, list), "Chunk data reader must return a `list(samples)`"
+            except StopIteration:
+                pass
+
+            # Updates internal cache
+            if new_chunk and cache_size > 0:
+                self._cache = self._cache + new_chunk
+            elif new_chunk:
+                self._cache = new_chunk
+            elif not new_chunk and cache_size == 0:
+                raise StopIteration
+
+            # Apply second level of shuffling
+            if self.shuffle_cache:
+                random.shuffle(self._cache)
```

Comment:
Nit: Should we shuffle the whole cache or just the new_chunk data? I am debating too as each has its own pros and cons. But I feel shuffle just the new_chunk is clearer and easier for repro in case of debugging?

Reply:
For debugging, the user should set `shuffle=False` for a predictable behavior, otherwise it will be hard to debug in either case :)

I will keep this open so that we can discuss the usefulness of both approach further

Reply:
@jaliyae @xzhu1900

Reply:
I agree with Xueyun on this. Shuffle should be the chunk not the whole cash. I think the code complexity is the same in both case, but if we shuffle the cache we cannot explain our algorithm easily.

Reply:
fixed